From 7a6018a823249acab3b36906594f14d16055dc32 Mon Sep 17 00:00:00 2001
From: Haohui Mai <wheat9@apache.org>
Date: Sun, 22 Nov 2015 23:07:43 -0800
Subject: [PATCH 1294/2848] HADOOP-12313. NPE in JvmPauseMonitor when calling
 stop() before start(). Contributed by Gabor
 Liptak.

(cherry picked from commit f826195ee541a6fce8d287ad1a49400ff2128c09)

Conflicts:
	hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/JvmPauseMonitor.java

Change-Id: I0f7ef2f85f5c0c56a5752c4c26f2baa6f34b02a3
---
 .../org/apache/hadoop/util/JvmPauseMonitor.java    |   18 ++++++++++++------
 1 file changed, 12 insertions(+), 6 deletions(-)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/JvmPauseMonitor.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/JvmPauseMonitor.java
index 33d46d8..72ccf9d 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/JvmPauseMonitor.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/JvmPauseMonitor.java
@@ -78,16 +78,22 @@ public void start() {
     Preconditions.checkState(monitorThread == null,
         "Already started");
     monitorThread = new Daemon(new Monitor());
-    monitorThread.start();
+    if (shouldRun) {
+      monitorThread.start();
+    } else {
+      LOG.warn("stop() was called before start() completed");
+    }
   }
   
   public void stop() {
     shouldRun = false;
-    monitorThread.interrupt();
-    try {
-      monitorThread.join();
-    } catch (InterruptedException e) {
-      Thread.currentThread().interrupt();
+    if (isStarted()) {
+      monitorThread.interrupt();
+      try {
+        monitorThread.join();
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      }
     }
   }
 
-- 
1.7.9.5

